### PR TITLE
Refactor of the submission utility

### LINKED ISSUE #232: Generalize status code handling
Per [this comment](https://github.com/elyra-ai/elyra/pull/231#pullrequestreview-339297287) we should probably relax some of the status code interpretation in order to better facilitate server-side changes or even other servers in general without having to update front-end code to handle the addition of success (and error) codes.

In particular, we should expand the set of success status codes from 200 (which is currently all that the elyra server returns) to also include 201 (when resources are created via POST), 204 (when resources are deleted), etc.

### LINKED ISSUE #271: Create a service class for runtimes connections
Create a service class like `CodeSnippetManager` added in #241 for handling calls to `api/metadata/runtimes`. The new class would be located in the `application` package for use by the pipeline editor and the button.

### LINKED ISSUE #400: Update submission util to be more generic
As discussed in #241 the new code-snippet package has brought to light issues with the submission utils in application.

copied from my comment in #241:
> The current util is too specific to runtimes and pipeline submissions and the errors read wrong for code snippets, relatedly I believe the issues causing us to use a signal is a self-created problem from using the existing submission util that assumes it's only called by a user taking an action, not as part of startup as code snippet uses it.

From that I believe we should make our error handling more generic, allowing more control over how errors are handled, like letting errors surface in ways other than a blocking dialog and making sure error text matches the error case.

### PR DESCRIPTION
**Pull request by ajbozarth**
The submission utility in application has become a mess since
it's creation due a limited view of it's usage at conception
and extending it's functionality over time.

This overhaul of the utility accomplishes the following:
- Moves from passing call back functions to using async utility functions
- Moves pipeline related utility functions into pipeline-editor (Part of #494)
- Update all calls to metadata end points to use the metadata service. Fixes #271
- Move ErrorDialogContent into ui-components. First step towards #403
- Update function, class, and file names to match actual functionality
- Modularized HTTP_STATUS success code verification. Fixes #232

Fixes #400

 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

**Comment by ajbozarth** [comment]
Sorry the diff is poor. Due to all the renamed files/extensions it is hard to tell what's changed in those files. This pr will also need to make some updates once #546 is merged since I plan to move `ErrorDialogContent` since this PR should be followed up with a refactor merging `ExpandableComponent` and `ErrorDialogContent` to finish #403

File: packages/application/src/services.tsx [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    return showDialog({
+      title: 'Error retrieving metadata',
+      body: <p>No {metadataName} metadata has been configured.</p>,
+      buttons: [Dialog.okButton()]
+    });
+  }
+
+  static async getMetadata(name: string): Promise<any> {
+    const metadataResponse: any = await RequestHandler.makeGetRequest(
+      'api/metadata/' + name,
```
Comment (kevin-bates) [inline packages/application/src/services.tsx:36]: `name` is a _namespace_ - we should name it as such.

File: packages/code-snippet/src/CodeSnippetManager.ts [hunk 2]
```diff
@@ -25,37 +25,25 @@ export interface ICodeSnippet {
...
+    const allCodeSnippets: ICodeSnippet[] = [];
+
+    for (const codeSnippetKey in codeSnippetsResponse) {
+      const jsonCodeSnippet = codeSnippetsResponse[codeSnippetKey];
+      const codeSnippet: ICodeSnippet = {
+        name: jsonCodeSnippet.name,
+        displayName: jsonCodeSnippet.display_name,
+        description: jsonCodeSnippet.metadata.description,
+        language: jsonCodeSnippet.metadata.language,
+        code: jsonCodeSnippet.metadata.code
```
Comment (kevin-bates) [inline packages/code-snippet/src/CodeSnippetManager.ts:41]: Hmm.  Should we have the server _model_ promote the various metadata-level properties to the same as `name` and `description`?  If I'm reading these correctly, you could then have `jsonCodeSnippet` implement `ICodeSnippet`.

File: packages/pipeline-editor/src/SubmissionHandler.tsx [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+import { showDialog, Dialog } from '@jupyterlab/apputils';
+import * as React from 'react';
+
+export class SubmissionHandler {
+  static async getRuntimeImages(): Promise<IDictionary<string>> {
+    const runtimeImages = await FrontendServices.getMetadata('runtime-images');
+    const images: { [key: string]: string } = {};
+    for (const image in runtimeImages) {
+      const imageName: string = runtimeImages[image]['metadata']['image_name'];
+      images[imageName] = runtimeImages[image]['display_name'];
```
Comment (kevin-bates) [inline packages/pipeline-editor/src/SubmissionHandler.tsx:31]: I think using the `name` field would probably make for a better index than `image_name`.  `name` is derived from the [instance's filename](https://github.com/elyra-ai/elyra/tree/master/etc/config/metadata/runtime-images), whereas `image_name` may contain ["unfriendly" characters](https://github.com/elyra-ai/elyra/blob/master/etc/config/metadata/runtime-images/tensorflow_2x_gpu_py3.json#L4).

File: packages/pipeline-editor/src/SubmitNotebook.ts [hunk 4]
```diff
@@ -73,52 +73,33 @@ export class SubmitNotebookButtonExtension
...
-
-              SubmissionHandler.submitPipeline(
-                pipeline,
-                result.value.runtime_config,
-                'notebook'
-              );
-            });
-          }
-        );
+    const runtimes = await FrontendServices.getMetadata('runtimes');
```
Comment (kevin-bates) [inline packages/pipeline-editor/src/SubmitNotebook.ts:81]: `getMetadata('runtimes')` is called from `handleExport()` and `handleRun()`, as well as here.  I can't tell if `FrontendServices.getMetadata()` unconditionally issues a request to the server or not, but its good that we have a central place in case we want to cache metadata in general.

File: packages/application/src/services.tsx (hunk 1 above)
Comment (ajbozarth) [reply packages/application/src/services.tsx:36]: @lresende wrote this in a previous pr but I agree with you so I'll update it

File: packages/code-snippet/src/CodeSnippetManager.ts (hunk 2 above)
Comment (ajbozarth) [reply packages/code-snippet/src/CodeSnippetManager.ts:41]: This was written by @karlaspuldaro originally with @lresende help. Iiuc I believe `ICodeSnippet` was written to capture what the UI currently needs, but the back end schema (what the api returns) was written with the expectation of adding more metadata in the future and includes fields we may not yet use in the UI.

File: packages/pipeline-editor/src/SubmissionHandler.tsx (hunk 3 above)
Comment (ajbozarth) [reply packages/pipeline-editor/src/SubmissionHandler.tsx:31]: Sadly we didn't get to pick the index. The json returned here is actually copied as is into a canvas schema. So it has to match the expected format :(

Comment (ajbozarth) [reply packages/pipeline-editor/src/SubmissionHandler.tsx:31]: addendum to my previous comment, we do have a function that inserts the json into that schema file so we could always make this cleaner and then "revert" is to the necessary format in that function. If you really feel the service function should return a better formatted dict.

File: packages/pipeline-editor/src/SubmitNotebook.ts (hunk 4 above)
Comment (ajbozarth) [reply packages/pipeline-editor/src/SubmitNotebook.ts:81]: The original intention of #271 was to create a service class for runtimes, so I initially started writing a function like `getRuntimeImages` but I stop when I realized that it would just be:
```
static async getRuntimes(): Promise<any> {
    const runtimes = await FrontendServices.getMetadata('runtimes');
    return runtimes;
}
```
and I felt that was silly. Unlike our other metadata runtimes returns already in the format we need to use it.

Since we currently don't cache metadata I didn't think it was worth it, but if we want to enable an easy way to do that in the future I can create the `getRuntimes()` function in SubmissionHandler to be used in those the there uses.

File: packages/application/src/services.tsx (hunk 1 above)
Comment (ajbozarth) [reply packages/application/src/services.tsx]: pushed an update for this

**Comment by ajbozarth** [comment]
@kevin-bates thanks for the quick review. As for the api endpoint changes, that should be easy to do in a follow up pr since with this pr we only write out each of the three endpoints once on the front end.

File: packages/code-snippet/src/CodeSnippetManager.ts (hunk 2 above)
Comment (kevin-bates) [reply packages/code-snippet/src/CodeSnippetManager.ts:41]: Ok - if there's no advantage to removing the `metadata` level, I'm good.  This is the kind of thing we need to do prior to 1.0.  I suppose there's some advantage to keeping the metadata-specific properties contained.

File: packages/pipeline-editor/src/SubmissionHandler.tsx (hunk 3 above)
Comment (kevin-bates) [reply packages/pipeline-editor/src/SubmissionHandler.tsx:31]: Could you please provide a reference to the schema file?  I'm sure it's from https://github.com/elyra-ai/pipeline-schemas but never know which is what in there.

Changing the index sounds like more effort than its worth since we don't really control things here.  I figured `images[]` was our thing.

File: packages/pipeline-editor/src/SubmitNotebook.ts (hunk 4 above)
Comment (kevin-bates) [reply packages/pipeline-editor/src/SubmitNotebook.ts:81]: Why wouldn't we just add the caching into `getMetadata` and allow all metadata namespaces to use the same caching framework - a cache per namespace - when that time comes?

**Comment by kevin-bates** [comment]
Correct - this is exactly why #513 is [pending the merge of this PR](https://github.com/elyra-ai/elyra/issues/513#issuecomment-628763087).

File: packages/application/src/services.tsx [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  }
+
+  static async getMetadata(namespace: string): Promise<any> {
+    const metadataResponse: any = await RequestHandler.makeGetRequest(
+      'api/metadata/' + namespace,
+      false
+    );
+
+    if (Object.keys(metadataResponse[namespace]).length === 0) {
+      return this.noMetadataError(namespace);
```
Comment (vabarbosa) [inline packages/application/src/services.tsx:41]: i don't think an empty list necessarily means an error retrieving metadata. for example, i could have a `code-snippets` namespace but just not have any code snippets at the moment. in this case an empty list is returned which i think should be valid and not considered an error. i think the error dialog should be presented only if the request returns an actual error.

in addition the use of a dialog here can become tedious. if i tend to leave the code-snippet panel open by default, each time i refresh or relaunch lab i am prompted with the dialog. instead of opening an error dialog return the empty list and leave it up to the caller to decide how to handle the empty list. for example, code snippets could just display some placeholder text (i.e., 'No code snippets have been configured') in the panel when there are no snippets.

File: packages/application/src/requests.tsx [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    requestOptions: any,
+    displayWaitDialog: boolean
+  ): Promise<any> {
+    // use ServerConnection utility to make calls to Jupyter Based services
+    // which in this case are the in the extension installed by this package
+    const settings = ServerConnection.makeSettings();
+    const requestUrl = URLExt.join(settings.baseUrl, requestExt);
+
+    console.log(
+      'Sending a ' + requestOptions.method + ' request to ' + requestUrl
```
Comment (vabarbosa) [inline packages/application/src/requests.tsx:97]: more of a nit (or personal preference on my part). i see this (string concatenation) throughout. any particular reason for not using template literals instead? i tend to find them more readable. for example

```javascript
console.log(`Sending a ${requestOptions.method} request to ${requestUrl}`)
```

File: packages/pipeline-editor/src/SubmissionHandler.tsx (hunk 3 above)
Comment (ajbozarth) [reply packages/pipeline-editor/src/SubmissionHandler.tsx:31]: This is the template we insert it into https://github.com/elyra-ai/elyra/blob/1947597c3b3e63e7377ac9bffae6f1c083f2fe9d/packages/pipeline-editor/src/pipeline-template.json

File: packages/application/src/services.tsx (hunk 5 above)
Comment (ajbozarth) [reply packages/application/src/services.tsx:41]: This is a good argument for having a separate `getRuntimes` function like @kevin-bates and I discussed in the other thread above. Then we could take this check out of the getMetadata call and move it into the specific namespace wrappers that require it (runtimes/runtimeImages). I'll make the changes

File: packages/application/src/requests.tsx (hunk 6 above)
Comment (ajbozarth) [reply packages/application/src/requests.tsx:97]: This was old code, but I agree, I'll update it

File: packages/pipeline-editor/src/SubmissionHandler.tsx (hunk 3 above)
Comment (kevin-bates) [reply packages/pipeline-editor/src/SubmissionHandler.tsx:31]: So the _keys_ of `images` wind up filling in the [`{{image}}`](https://github.com/elyra-ai/elyra/blob/1947597c3b3e63e7377ac9bffae6f1c083f2fe9d/packages/pipeline-editor/src/pipeline-template.json#L16) in the template?

And does `display_name` get used to map to the image in a dropdown somewhere?

Comment (ajbozarth) [reply packages/pipeline-editor/src/SubmissionHandler.tsx:31]: whoops that was the wrong file https://github.com/elyra-ai/elyra/blob/87f864b6951c1fc5281ccfd9166e7f5945098825/packages/pipeline-editor/src/properties.json

File: packages/application/src/requests.tsx [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+              resolve(result);
+            },
+            (reason: any) => {
+              // handle 404 if elyra server extension is not found
+              return this.server404();
+            }
+          );
+        }
+      );
+    });
```
Comment (karlaspuldaro) [inline packages/application/src/requests.tsx:132]: I find the flow here confusing, and there might be exceptions not being caught, then serverResponse could never be resolved. I believe at the end of every possible end point should be either a resolve or reject, instead of returns.
- ServerConnection.makeRequest doesn't have a catch for promise rejection
- if response.json fails, does it fall into this.server404? 
- if response.json gets a result with a server error status, shouldn't line 120 resolve the promise instead of using return?

File: packages/application/src/ErrorDialogContent.tsx [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+    const width = this.dialogNode.clientWidth;
+    const height = this.dialogNode.clientHeight;
+
+    if (
+      expanded &&
+      (width < ERROR_DIALOG_WIDTH || height < ERROR_DIALOG_HEIGHT)
+    ) {
+      this.dialogNode.style.width = ERROR_DIALOG_WIDTH + 'px';
+      this.dialogNode.style.height = ERROR_DIALOG_HEIGHT + 'px';
```
[the remark this answers is not in the document]
Comment (ajbozarth) [reply packages/application/src/ErrorDialogContent.tsx:62]: @vabarbosa could explain better, but I believe it's a bit of a hack and it wouldn't work using plain css

[the remark this answers is not in the document]
Comment (vabarbosa) [reply packages/application/src/ErrorDialogContent.tsx:62]: good question. it a result of the way [CSS resize](https://developer.mozilla.org/en-US/docs/Web/CSS/resize) works.

when an HTML element is resizeable browsers will explicitly set its the width and height attributes inline as it's being resized. therefore using CSS classes for width and height would not work as the inline style would supersede it, so the way to reset/change the element's size is by updating the inline style.

File: packages/application/src/requests.tsx [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+import { showDialog, Dialog } from '@jupyterlab/apputils';
+import { URLExt } from '@jupyterlab/coreutils';
+import { ServerConnection } from '@jupyterlab/services';
+
+import * as React from 'react';
+
+import { ErrorDialogContent } from './ErrorDialogContent';
+
+const HTML_STATUS_SUCCESS = [200, 201];
```
Comment (lresende) [inline packages/application/src/requests.tsx:25]: HTTP response status codes indicate whether a specific HTTP request has been successfully completed. Responses are grouped in five classes:

- Informational responses (100–199),
- Successful responses (200–299),
- Redirects (300–399),
- Client errors (400–499),
- Server errors (500–599).


Within successful responses these are defined on the RFC 

- 200 ok
- 201 created
- 202 accepted
- 203 Non-Authoritative information
- 204 No Content
- 205 Reset Content
- 206 Partial Content
and a few other WebDAV related

File: packages/application/src/requests.tsx [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+export class RequestHandler {
```
Comment (lresende) [inline packages/application/src/requests.tsx:27]: Can we make this visible only to application package, and expose this via Application.invokeService or something like that ?

File: packages/application/src/requests.tsx [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+
+  static server404(): Promise<Dialog.IResult<any>> {
+    return showDialog({
+      title: 'Error contacting server',
+      body: <p>Elyra service endpoint not found.</p>,
+      buttons: [Dialog.okButton()]
+    });
+  }
+
+  static async makeGetRequest(
```
Comment (lresende) [inline packages/application/src/requests.tsx:63]: I believe that, with the latest @kevin-bates updates, we need to perform all the CRUD operations. Is the expectation that other type of requests fallback to makeServerRequest?

File: packages/application/src/requests.tsx [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+    displayWaitDialog: boolean
+  ): Promise<any> {
+    return this.makeServerRequest(
+      requestExt,
+      { method: 'POST', body: requestBody },
+      displayWaitDialog
+    );
+  }
+
+  static async makeServerRequest(
```
Comment (lresende) [inline packages/application/src/requests.tsx:86]: we can use some documentation here on the parameters

File: packages/application/src/requests.tsx [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+    requestExt: string,
+    requestOptions: any,
+    displayWaitDialog: boolean
```
Comment (lresende) [inline packages/application/src/requests.tsx:89]: Could we have a separate request, like `makeServerLongRequest` that would set this as true. We could then implement things like timeout, retries, etc?

File: packages/application/src/services.tsx [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+
+import { Dialog, showDialog } from '@jupyterlab/apputils';
+import * as React from 'react';
+
+import { RequestHandler } from './requests';
+
+/**
+ * A utilities class for various elyra services.
+ */
+export class FrontendServices {
```
Comment (lresende) [inline packages/application/src/services.tsx:25]: Jupyter already has `JupyterFrontEnd`  class which I really think that it wasn't a good name choice, and with this new one, with a similar name, could confuse users? What about just calling this like `Application` and then we would have `Application.getMetadata()`?

File: packages/application/src/services.tsx [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+    });
+  }
+
+  static async getMetadata(namespace: string): Promise<any> {
+    const metadataResponse: any = await RequestHandler.makeGetRequest(
+      'api/metadata/' + namespace,
+      false
+    );
+
+    return metadataResponse[namespace];
```
Comment (lresende) [inline packages/application/src/services.tsx:40]: Is the dict format returned here the best option to return, or should we use a list?
Understanding the usage pattern by the UI components would help on this decision (e.g if all consumers loop trough and populate UI components, list would probably make more sense... if there are scenarios where lookups are used, then dict might actually be the best option. Another approach is to enable something like getMetadata().asDict and getMetadata.asList()

File: packages/pipeline-editor/src/SubmissionHandler.tsx [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { FrontendServices, RequestHandler } from '@elyra/application';
+import { showDialog, Dialog } from '@jupyterlab/apputils';
+import * as React from 'react';
+
+export class SubmissionHandler {
```
Comment (lresende) [inline packages/pipeline-editor/src/SubmissionHandler.tsx:21]: Again, this looks good, but it would be good to actually have a facade service here as well, with something like PipelineService.submitPipeline()which will somewhat follow the pattern we are trying to establish throughout the project.

File: packages/application/src/ErrorDialogContent.tsx [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+
+import { caretDownIcon, caretRightIcon } from '@jupyterlab/ui-components';
+
+import * as React from 'react';
+
+const MESSAGE_DISPLAY = 'elyra-pipelineSubmission-messageDisplay';
+const ERROR_DISPLAY_BUTTON = 'elyra-pipelineSubmission-errDisplayButton';
+const ERROR_DETAILS = 'elyra-pipelineSubmission-errDetails';
+const ERROR_DETAILS_VISIBLE = 'elyra-pipelineSubmission-error-visible';
+const ERROR_DETAILS_HIDDEN = 'elyra-pipelineSubmission-error-hidden';
```
[the remark this answers is not in the document]
Comment (ajbozarth) [reply packages/application/src/ErrorDialogContent.tsx:25]: nice catch, that actually just poor naming since the function that utilizes it used to be called `pipelineSubmission`. I'll update the css names to be generic

File: packages/application/src/requests.tsx (hunk 9 above)
Comment (ajbozarth) [reply packages/application/src/requests.tsx:25]: I figured we could add more valid cases here as we have them. Currently these are the only two valid success statuses that the elyra api returns. I can make this check for any value in the 200–299 range instead, but I though it'd be better to be specific because if we get a "success" code that shouldn't be possible we should treat it as an error.

File: packages/application/src/requests.tsx (hunk 10 above)
Comment (ajbozarth) [reply packages/application/src/requests.tsx:27]: But this is meant to be a utility class, though I would agree that making all but `makePostRequest` and `makeGetRequest` private would be prudent. I'm unsure how we would make these two functions available outside of application otherwise.

File: packages/application/src/requests.tsx (hunk 12 above)
Comment (ajbozarth) [reply packages/application/src/requests.tsx:86]: will do, after talking to @karlaspuldaro I plan to add more documentation to this file in general

File: packages/application/src/requests.tsx (hunk 13 above)
Comment (ajbozarth) [reply packages/application/src/requests.tsx:89]: I'm not against this but how exactly would `makeServerLongRequest` be called. currently this function is only called by `makePostRequest` and `makeGetRequest` so would we need to add a `long` param to them?

File: packages/application/src/services.tsx (hunk 15 above)
Comment (ajbozarth) [reply packages/application/src/services.tsx:40]: This currently just returns the metadata "as is" and leaves the formatting to the service functions that call it. Each service function handles error cases and formatting to match how that metadata namespace is used by the frontend.

File: packages/application/src/services.tsx (hunk 14 above)
Comment (ajbozarth) [reply packages/application/src/services.tsx:25]: I also am not a fan our this name, but I don't think Application is any better. I's be more in favor of just naming it `MetadataService` since that's what it is.

File: packages/pipeline-editor/src/SubmissionHandler.tsx (hunk 16 above)
Comment (ajbozarth) [reply packages/pipeline-editor/src/SubmissionHandler.tsx:21]: I'll update this to be `PipelineService` I almost named it that during development anyway.

File: packages/application/src/requests.tsx (hunk 11 above)
Comment (ajbozarth) [reply packages/application/src/requests.tsx:63]: that was my original intention yes

File: packages/application/src/requests.tsx (hunk 13 above)
Comment (ajbozarth) [reply packages/application/src/requests.tsx:89]: thinking about this more I'm unsure how creating a wrapper function around this setting this to true would enable things like timeout, retries, etc. `makeServerRequest` handles any "errors" itself. If we wanted it to be more extendable we would have to let `makeServerRequest` return errors and add error handling to every service that calls it, most of which would be duplicate code that currently resides in `makeServerRequest`

File: packages/application/src/requests.tsx (hunk 7 above)
Comment (ajbozarth) [reply packages/application/src/requests.tsx:130]: @karlaspuldaro and I discussed this in detail over Webex to get both our head around how it work and I will be adding in-line documentation to explain it, no code changes are required.

Comment (ajbozarth) [reply packages/application/src/requests.tsx]: added comments

File: packages/application/src/requests.tsx (hunk 12 above)
Comment (ajbozarth) [reply packages/application/src/requests.tsx:123]: added doc comments

File: packages/pipeline-editor/src/SubmissionHandler.tsx (hunk 16 above)
Comment (ajbozarth) [reply packages/pipeline-editor/src/SubmissionHandler.tsx]: renamed

**Comment by ajbozarth** [comment]
@lresende I've pushed some updates address some of your comments as well updating to latest master and moving ErrorDialogContent.tsx to ui-components.

File: packages/application/src/requests.tsx (hunk 6 above)
Comment (lresende) [reply packages/application/src/requests.tsx]: @vabarbosa Please definetlly educate the `backend engineers`

File: packages/application/src/requests.tsx (hunk 9 above)
Comment (lresende) [reply packages/application/src/requests.tsx:24]: I would prefer that the accepted range be (200-299) but would be ok with (200-2010) for now. @kevin-bates thoughts here? what are the success status code range you have been using?

File: packages/application/src/services.tsx (hunk 15 above)
Comment (lresende) [reply packages/application/src/services.tsx:40]: Yes, the "as is" is a dict. And then all the consumers have to convert that to list, at least on couple places I saw.

File: packages/application/src/requests.tsx [hunk 18]
```diff
@@ -82,7 +109,18 @@ export class RequestHandler {
       displayWaitDialog
     );
   }
-
+  /**
+   * Make an request to the jupyterlab server.
+   *
+   * @param requestExt - The url for the request.
+   *
```
Comment (lresende) [inline packages/application/src/requests.tsx:115]: the `Ext` here really made me confused, could we make a better name here? maybe requestURL, endpoint, servicePath...

File: packages/application/src/requests.tsx [hunk 19]
```diff
@@ -82,7 +109,18 @@ export class RequestHandler {
...
+   * @param requestOptions - The initialization options for the request.
+   *
```
Comment (lresende) [inline packages/application/src/requests.tsx:117]: What are these options? Valid values?

File: packages/application/src/requests.tsx [hunk 20]
```diff
@@ -82,7 +109,18 @@ export class RequestHandler {
...
+   * @param displayWaitDialog - Whether or not to display a dialog warning
+   * warning the use the request may take time.
+   *
```
Comment (lresende) [inline packages/application/src/requests.tsx:120]: If we don't make a wrapper, could we make this a more generic name that flags the call as expensive, which will take a while to be submitted, and would for now require the dialog to show up, but could enable other things to be handled in this special case?

File: packages/application/src/services.tsx [hunk 21]
```diff
@@ -20,9 +20,9 @@ import * as React from 'react';
 import { RequestHandler } from './requests';
 
 /**
- * A utilities class for various elyra services.
+ * A utility class for handling elyra metadata calls.
  */
-export class FrontendServices {
+export class MetadataService {
```
Comment (lresende) [inline packages/application/src/services.tsx:25]: This should be a facade for multiple common things available from the Application package and usable for multiple Elyra packages, so metadata service is not a good name. I rather stay with FrontendServices if we are not going to use Application

File: packages/application/src/requests.tsx (hunk 9 above)
Comment (kevin-bates) [reply packages/application/src/requests.tsx:24]: We're only using 200 and 201.  204 (No Content) could be used on delete if we were to choose not to return the deleted resource, but since we return content we use 200 on deletes.

File: packages/application/src/services.tsx (hunk 15 above)
Comment (kevin-bates) [reply packages/application/src/services.tsx:40]: If this call is just calling `GET /api/metadata/{namespace}` we return a JSON object of the form: [`{"runtimes": [ array of metadata instances ]}`](https://github.com/elyra-ai/elyra/blob/master/elyra/api/elyra.yaml#L126-L134), so it's both (a dict and a list).

File: packages/application/src/requests.tsx (hunk 7 above)
Comment (karlaspuldaro) [reply packages/application/src/requests.tsx]: I still think this is an important piece since it's being called by any get and post requests, and as agreed in our discussion, it at least requires more testing for edge cases to make sure we don't have problems maintaining this in the future. 

I tried running  the prototype below in node to see what happens when something goes wrong somewhere inside, await doesn't seem to catch all exceptions if they're not properly handled:
```
const asyncFunc = async () => {
	// First promise
	const p1 = new Promise((res, rej) => {
		
		// Second promise
		new Promise((res2,  rej2) => { 
		setTimeout(
			() => {
				// Let's say there is some sort of error in this inner promise
				console.log("rejecting second promise");
				rej2("exception: rej2");
			}
			, 1000)}
		)
		.then(
			result => {
				if(!2===2){
					return console.log("some server error");
					// why not resolve
				}
				resolve(result);
			},
			error =>  {
				// Exception catch block
				console.log("404");
				// do we want to resolve or reject here?
			}
		);

	}); // no exception catch block for p1
	

	const result = await p1;
	console.log("p1 done! p1 res = "  + result + "\nReturning...");
	return result;
}


asyncFunc();
console.log("waiting for promise...");
```
And my output was:

```
waiting for promise...
rejecting second promise
404
```
In this case, makeServerRequest never gets resolved and its return is never called, even if it called 404 callback and the error bubbled up.
Even if we only treat the specific cases we want our flow to fall into and not expose anything unnecessary to the client, maybe we should make sure every possible promise path is being handled.
I'm not totally sure what are the consequences of hanging promises, or even if our flow is somehow controlled and changes are not required, test scenarios can answer this question.

Comment (ajbozarth) [reply packages/application/src/requests.tsx]: ok so your argument is that we should "handle" all the possible error cases, even if we should reach them. (I would just raise an error of some sort in those cases, not a dialog)

File: packages/application/src/requests.tsx (hunk 9 above)
Comment (ajbozarth) [reply packages/application/src/requests.tsx:24]: though I would normally want to only handle expected cases as success, since this is a util I will update this to handle all 200-299 as success to follow html standard

File: packages/application/src/requests.tsx (hunk 18 above)
Comment (ajbozarth) [reply packages/application/src/requests.tsx:115]: this part of the url is traditionally refereed to as the `path` so how about I update it to `requestPath`?

File: packages/application/src/requests.tsx (hunk 19 above)
Comment (ajbozarth) [reply packages/application/src/requests.tsx:117]: this is passed as is to `makeRequest` but that function calls it `init` and defines it as `The initialization options for the request` should we change our wording to match?

File: packages/application/src/requests.tsx (hunk 20 above)
Comment (ajbozarth) [reply packages/application/src/requests.tsx:120]: hmmm, I guess we can. I'll change the name to `longRequest`, it's more generic but still gets its purpose across.

File: packages/application/src/services.tsx (hunk 21 above)
Comment (ajbozarth) [reply packages/application/src/services.tsx:25]: Ok, I'll change it back to `FrontendServices` then, though if you prefer `ApplicationServices` I believe I would be ok with that. I just want to the classname to be clear it's a "services" class

**Comment by ajbozarth** [comment]
@lresende @kevin-bates @karlaspuldaro I believe I've addressed everyones's review comments and managed to "rebase" on master (the commit history is a mess, but that'll be fixed by a squash before merge anyway). 

I also should note that a few open PRs will step on the toes of this one and will require updates depending on what merges first: #544 #560 #561

**Comment by ajbozarth** [comment]
Once #561 is merged I'll update this to address the changes in #544 #560 and #561

**Comment by ajbozarth** [comment]
I just pushed an update that merges in the latest master and addresses some issues raised in the merge. This should be good to go now if everyone could take a look and review

File: packages/application/src/requests.tsx (hunk 19 above)
Comment (lresende) [reply packages/application/src/requests.tsx:113]: As a consumer, how do I figure out what valid options that I should use here? Is this based on some other function parameter field requirement/definition? is there any documentation available to help consumers?

File: packages/application/src/requests.tsx (hunk 12 above)
Comment (lresende) [reply packages/application/src/requests.tsx:119]: I think more documentation is good, but I am really asking for [typedoc](https://typedoc.org/guides/doccomments/) for the function objective and expected inputs/outputs

File: packages/ui-components/src/ErrorDialogContent.tsx [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+
+interface IErrorDialogProps {
+  reason: string;
+  message: string;
+  timestamp: string;
+  traceback: string;
+  default_msg: string;
+}
+
+export class ErrorDialogContent extends React.Component<
```
Comment (lresende) [inline packages/ui-components/src/ErrorDialogContent.tsx:38]: Is this more like an `ErrorDialogWidget` or `ErrorDialogComponent`

File: packages/pipeline-editor/src/PipelineService.tsx [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+      pipeline: pipeline,
+      export_format: pipeline_export_format,
+      export_path: pipeline_export_path,
+      overwrite: overwrite
+    };
+
+    await RequestHandler.makePostRequest(
+      'api/pipeline/export',
+      JSON.stringify(body),
+      true
```
Comment (lresende) [inline packages/pipeline-editor/src/PipelineService.tsx:118]: this should be false, we don't have to create/upload any dependencies etc when exporting

File: packages/ui-components/src/ErrorDialogContent.tsx (hunk 22 above)
Comment (ajbozarth) [reply packages/ui-components/src/ErrorDialogContent.tsx:38]: I just kept the name, but following our other dialog body naming conventions we could change this to just `ErrorDialog`, `ServerErrorDialog`, or ExpandableErrorDialog` depending on how we want to classify it. I think the last one would be the most verbose and clear

File: packages/pipeline-editor/src/PipelineService.tsx (hunk 23 above)
Comment (ajbozarth) [reply packages/pipeline-editor/src/PipelineService.tsx:118]: I can make this `false` if that's what you want, I was just keeping current functionality, which is that exporting also displayed the dialog. As far as my experience, exporting still took enough time that it was worth having the dialog though.

**Comment by lresende** [comment]
I think I am done with my review, and I would like a few things before approving:
* Name consistency: PipelineService.ts and CodeSnippetService.ts
* Updated documentation on services exposed by Application package

Other things could be handled in future prs.

File: packages/code-snippet/src/CodeSnippetWidget.tsx [hunk 24]
```diff
@@ -40,7 +39,7 @@ import { Widget } from '@lumino/widgets';
 
 import React from 'react';
 
-import { CodeSnippetManager, ICodeSnippet } from './CodeSnippet';
+import { CodeSnippetManager, ICodeSnippet } from './CodeSnippetManager';
```
[the remark this answers is not in the document]
Comment (ajbozarth) [reply packages/code-snippet/src/CodeSnippetWidget.tsx]: I already fixed this, and whether a service file is ts or tsx would only depend on whether it has to leverage react in it. Some do, some don't.
